### PR TITLE
Add a different implementation of Differentiation

### PR DESCRIPTION
**Pull request by edymtt**
In the long term, the tgmath derivatives currently implemented in
Differentation will be migrated to a separate overlay.

In anticipation of this, provide an implementation of Differentation
without such functions which is exposed through a feature flag.

Addresses rdar://68471851

### CODE REVIEW

**Comment by edymtt** [comment]
@swift-ci please smoke test

**Review by dan-zheng — APPROVED** [review]

**Comment by rxwei** [comment]
Note for reviewers: Since differentiation is platform-independent, it shouldn't need to depend on platform runtime libraries like Darwin and Glibc. One possible future direction is to move tgmath derivatives to a [cross-import overlay](https://forums.swift.org/t/cross-import-overlays/30868/) for Darwin/Glibc and _Differentiation.

**Review by rxwei — APPROVED** [review]

File: CMakeLists.txt [hunk 1]
```diff
@@ -402,6 +402,10 @@ option(SWIFT_ENABLE_EXPERIMENTAL_DIFFERENTIABLE_PROGRAMMING
   "Enable experimental Swift differentiable programming features"
   FALSE)
 
+option(SWIFT_EXPERIMENTAL_DIFFERENTIABLE_PROGRAMMING_FUTURE_IMPLEMENTATION
```
Comment (compnerd) [inline CMakeLists.txt:405]: @edymtt what do you think of sinking this option to `stdlib/CMakeLists.txt` since it doesn't impact the compiler at all?

Comment (edymtt) [reply CMakeLists.txt:405]: That's a good idea -- I missed that there were some options in that file as well

**Comment by edymtt** [comment]
@swift-ci please smoke test

**Comment by edymtt** [comment]
@swift-ci please test Windows

File: CMakeLists.txt (hunk 1 above)
Comment (edymtt) [reply CMakeLists.txt]: Should be addressed in b55f31bd98e9f00dea8d4e418614ce7925fdc367

**Comment by edymtt** [comment]
In addition to moving the option at the stdlib level, I also took some offline feedback from @gottesmm and moved the new implementation to a separate `CMakeLists.txt` file -- this should make the diff clearer and the code more readable overall.

**Review by rxwei — APPROVED** [review]

**Review by compnerd — APPROVED** [review]
Thanks!

**Comment by edymtt** [comment]
@swift-ci please smoke test macOS

**Comment by edymtt** [comment]
@swift-ci please smoke test macOS

File: stdlib/public/Differentiation_Future/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,20 @@
+set(SOURCES_FOLDER ../Differentiation)
```
Comment (compnerd) [inline stdlib/public/Differentiation_Future/CMakeLists.txt:1]: This seems in conflict with the comment in the parent CMakeLists, that this code should not be shared.

Comment (edymtt) [reply stdlib/public/Differentiation_Future/CMakeLists.txt:1]: It is indeed confusing -- my concern was around the CMake code, but the comment is not clear in this regard

I will sit a bit on the PR to see if I can improve the wording and/or the overall readability of the code

Comment (edymtt) [reply stdlib/public/Differentiation_Future/CMakeLists.txt:1]: I went ahead and did some rewording -- for the sake of clarify, did that in a new PR, https://github.com/apple/swift/pull/34116

**Comment by edymtt** [comment]
To avoid piling up many commits and to address properly the merge conflicts, close this PR in favor of

https://github.com/apple/swift/pull/34116

**Pull request closed without merging**